data_steward/analytics/cdr_ops/incremental_load_qc.py

```diff
@@ -0,0 +1,502 @@
+# we must pull these columns from `source_dataset`, not from `incremental_dataset`.<br>
+# If this QC fails, see how QC 4-2 goes.
+
+# +
+if not is_deid_dataset(new_dataset):
+    success_msg = "Skipping this check person_ext table exists only in DEID datasets.<br>"
+    render_message('', success_msg=success_msg)
+
+else:
+    query = JINJA_ENV.from_string('''
+        SELECT *
+        FROM `{{project}}.{{new_dataset}}.person_ext` n
+        JOIN `{{project}}.{{source_dataset}}.person_ext` s
+        ON n.person_id = s.person_id
+        WHERE (
+            n.state_of_residence_concept_id != s.state_of_residence_concept_id
+            OR n.state_of_residence_source_value != s.state_of_residence_source_value
+        )
+        OR (
+            (n.state_of_residence_concept_id IS NULL AND s.state_of_residence_concept_id IS NOT NULL)
+            OR (n.state_of_residence_source_value IS NULL AND s.state_of_residence_source_value IS NOT NULL)
+        )
+    ''').render(project=project_id,
+                new_dataset=new_dataset,
+                source_dataset=source_dataset)
+
+    df = execute(client, query)
+
+    success_null_check = (
+        f"All state columns in {new_dataset}.person_ext come from {source_dataset}, "
+        f"not {incremental_dataset}.<br><br>")
+    failure_null_check = (
+        f"There are <b>{len(df)}</b> records in {new_dataset}.person_ext that have "
+        f"unmatching state columns from {source_dataset}. <br>Look at the table and "
+        "investigate why they are inconsistent. <br><b>QC 4-2 will help narrow down "
+        "the potential cause of the issue.</b><br><br>")
+
+    render_message(df, success_null_check, failure_null_check)
+
+# -
+
+# ## QC 4-2. (Only for `is_deidentified == True`) Confirm `person_ext`'s state related columns come from `source_dataset` <br>- Check against records from `incremental_dataset`
+# If QC 4-1 fails, this QC will be helpful for troubleshooting. This QC checks the same thing as 4-1, but the scope is limited to only the participants from `incremental_dataset`.<br>
+# If QC 4-1 fails but 4-2 is successful, that means the issue comes from the original dataset, not from the incremental dataset. If both 4-1 and 4-2 fail, that means
+# something related to the incremental dataset did not work as expected.
+
+# +
+if not is_deid_dataset(new_dataset):
+    success_msg = "Skipping this check person_ext table exists only in DEID datasets.<br>"
+    render_message('', success_msg=success_msg)
+
+else:
+    query = JINJA_ENV.from_string('''
+        SELECT *
+        FROM `{{project}}.{{new_dataset}}.person_ext` n
+        JOIN `{{project}}.{{source_dataset}}.person_ext` s
+        ON n.person_id = s.person_id
+        WHERE n.person_id IN (
+            SELECT person_id FROM `{{project}}.{{incremental_dataset}}.person`
+            WHERE person_id IS NOT NULL
+        )
+        AND (
```

Comment:
Hi, @hiro-mishima - Thanks for adding this second check here.  I believe you need an additional wrapping set of parentheses around lines 283 to 290 to pair the OR condition with only the AND for state_of_residence check.

Reply:
Good catch, thanks @molly-giddings. I corrected the script with this commit: https://github.com/all-of-us/curation/pull/1489/commits/886e0229395bd04af7b63060acf39e77d297af23

Reply:
Thank you!  So speedy!